core/src/test/java/com/datastax/oss/driver/internal/core/ContactPointsTest.java

```diff
@@ -0,0 +1,177 @@
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableSet;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.Collections;
+import java.util.Set;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+import org.slf4j.LoggerFactory;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ContactPointsTest {
+
+  @Mock private Appender<ILoggingEvent> appender;
+  @Captor private ArgumentCaptor<ILoggingEvent> loggingEventCaptor;
+  private Logger logger;
+  private Level initialLogLevel;
+
+  @Before
+  public void setup() {
+    logger = (Logger) LoggerFactory.getLogger(ContactPoints.class);
+    initialLogLevel = logger.getLevel();
+    logger.setLevel(Level.INFO);
+    logger.addAppender(appender);
+  }
+
+  @After
+  public void teardown() {
+    logger.detachAppender(appender);
+    logger.setLevel(initialLogLevel);
+  }
+
+  @Test
+  public void should_parse_ipv4_address_and_port_in_configuration() {
+    Set<EndPoint> endPoints =
+        ContactPoints.merge(Collections.emptySet(), ImmutableList.of("127.0.0.1:9042"), true);
+
+    assertThat(endPoints)
+        .containsExactly(new DefaultEndPoint(new InetSocketAddress("127.0.0.1", 9042)));
+  }
+
+  @Test
+  public void should_parse_ipv6_address_and_port_in_configuration() {
+    Set<EndPoint> endPoints =
+        ContactPoints.merge(Collections.emptySet(), ImmutableList.of("0:0:0:0:0:0:0:1:9042"), true);
+
+    assertThat(endPoints).containsExactly(new DefaultEndPoint(new InetSocketAddress("::1", 9042)));
+  }
+
+  @Test
+  public void should_parse_host_and_port_in_configuration_and_create_unresolved() {
+    Set<EndPoint> endPoints =
+        ContactPoints.merge(Collections.emptySet(), ImmutableList.of("localhost:9042"), false);
+
+    assertThat(endPoints)
+        .containsExactly(
+            new DefaultEndPoint(InetSocketAddress.createUnresolved("localhost", 9042)));
+  }
+
+  @Test
+  public void should_parse_host_and_port_and_resolve_all_a_records() throws UnknownHostException {
+    int localhostARecordsCount = InetAddress.getAllByName("localhost").length;
+    assumeTrue(
+        "This test assumes that localhost resolves to multiple A-records",
+        localhostARecordsCount >= 2);
+
+    Set<EndPoint> endPoints =
+        ContactPoints.merge(Collections.emptySet(), ImmutableList.of("localhost:9042"), true);
+
+    assertThat(endPoints).hasSize(localhostARecordsCount);
+    assertLog(
+        Level.INFO,
+        "Contact point localhost:9042 resolves to multiple addresses, will use them all");
+  }
+
+  @Test
+  public void should_ignore_malformed_host_and_port_and_warn() {
+    Set<EndPoint> endPoints =
+        ContactPoints.merge(Collections.emptySet(), ImmutableList.of("foobar"), true);
```

Comment:
I think that we can create a parameterized test for `merge` that covers all cases

Reply:
I gave it a quick try, but I think it's not worth it. The parameters would be programmatic contact points, configuration contact points, whether to resolve, expected resulting endpoints, expected warnings:

```java
  @DataProvider
  public static Object[][] mergeScenarios() {
    final boolean resolve = true, doNotResolve = false; // for readability
    return new Object[][] {
      {
        Collections.emptySet(),
        ImmutableList.of("127.0.0.1:9042"),
        resolve,
        ImmutableSet.of(new DefaultEndPoint(new InetSocketAddress("127.0.0.1", 9042))),
        Collections.emptyList()
      },
    };
  }
```

That's harder to read than plain code, and not shorter. So considering that there are only a handful of tests in this class, it's better to favor clarity I think.